smarts/core/controllers/trajectory_interpolation_controller.py

```diff
@@ -0,0 +1,122 @@
+    TrajectoryTrackingControllerState,
+    TrajectoryTrackingController,
+)
+
+TIME_INDEX = 0
+X_INDEX = 1
+Y_INDEX = 2
+THETA_INDEX = 3
+VEL_INDEX = 4
+
+
+class TrajectoryInterpolationController:
+    @staticmethod
+    def ndarray_has_nan(ndarray: np.ndarray):
+        np.isnan(np.sum(ndarray))
+
+    @staticmethod
+    def interpolate(ms0: np.ndarray, ms1: np.ndarray, time: float) -> np.ndarray:
+        """Linear Interpolate between two vehicle motion state
+
+        Returns:
+            np.ndarray: New vehicle state between vehicle motion state ms0 and ms1
+        """
+
+        start_time = ms0[TIME_INDEX]
+        end_time = ms1[TIME_INDEX]
+        ratio = math.fabs((time - start_time) / (end_time - start_time))
+        assert end_time >= start_time and time >= start_time
+
+        np_motion_state = (1 - ratio) * ms0 + ratio * ms1
+        return np_motion_state
+
+    @staticmethod
+    def locate_motion_state(trajectory, time) -> np.ndarray:
+
+        if math.isinf(trajectory[TIME_INDEX][0]):
+            return trajectory[:, 0], trajectory[:, 0]
+
+        end_index = 0
+        for i, t in enumerate(trajectory[TIME_INDEX]):
+            if t > time:
+                end_index = i
+                break
+
+        if end_index <= 0:
+            raise RuntimeError(
+                "Next step time is %f, but cant be located at input traj" % time
+            )
+
+        return trajectory[:, end_index - 1], trajectory[:, end_index]
+
+    @classmethod
+    def perform_trajectory_interpolation(
+        cls, sim, agent_id, vehicle, trajectory: np.ndarray, controller_state,
+    ):
+        """Move vehicle with trajectory interpolation
+
+        If you want vehicle stop at a specific pose,
+        trajectory[TIME_INDEX][0] should be set as infinite
+
+        Args:
+            sim : reference of smarts instance
+            agent_id : agent who use this controller
+            vehicle : vehicle to be controlled
+            trajectory (np.ndarray): trajectory with time
+            controller_state : inner state of controller
+
+        Raises:
+            ValueError: Error when input trajectory has nan field.
+            RuntimeError: Error when input trajectory is empty
+        """
+        assert isinstance(vehicle.chassis, AckermannChassis)
+        if len(trajectory[TIME_INDEX]) <= 0:
+            raise RuntimeError("Input trajectory is empty!")
+        if cls.ndarray_has_nan(trajectory):
+            raise ValueError("Has nan in trajectory")
+
+        ms0, ms1 = cls.locate_motion_state(trajectory, sim.timestep_sec)
+
+        speed = 0.0
+        pose = []
+        if math.isinf(ms0[TIME_INDEX]) or math.isinf(ms1[TIME_INDEX]):
+            center_position = ms0[X_INDEX : Y_INDEX + 1]
+            center_heading = Heading(ms0[THETA_INDEX])
+            pose = Pose.from_center(center_position, center_heading)
+            speed = 0.0
+        else:
+            ms = cls.interpolate(ms0, ms1, sim.timestep_sec)
```

Comment:
Just wondering how you got this to work since `sim.timestep_sec` does not change?

```suggestion
            ms = cls.interpolate(ms0, ms1, sim._elapsed_sim_time)
```

Reply:
Yes, time of trajectory point should be relative. In my experience, trajectory should be updated at every planning cycle. Therefore, it is common to give trajectory point a relative time.

Reply:
OK, I was not sure how it was used but this should be fine then.

Reply:
So from my understanding, the main loop in the `locate_motion_state` finds the trajectory point near the next time step considering the case of mismatch between the temporal resolution of the trajectory and the time step of SMARTS, and using the obtained trajectory point and the point before it  from `locate_motion_state` we perform the interpolation for the amount of `sim.timestep_sec`. If my understanding is true, then this is ok for the case that the trajectory resolution is bigger than smarts time step,  however, for the case that the temporal resolution of the trajectory is smaller than the `sim.timestep_sec` this may result in significant error, in this case we might use an assert to ensure the temporal resolution relation or modify the return values of the `locate_motion_state`

Reply:
@iman512003 Your understanding is right. But I think it still be fine when the temporal resolution of the trajectory is smaller than the time step. I will submit some examples and tests to ensure the validity.